### DISCUSSION TITLE
Function alignment

### DISCUSSION THREAD

**Discussion by ethanaobrien** [emscripten-core/emscripten#22369] [Q&A]
I'm working on compiling [a certain emulator](https://github.com/libretro/picodrive) via emscripten, where [the lowest bit of every code pointer needs to be zero](https://github.com/libretro/picodrive/issues/233#issuecomment-2287223544).

Setting `-falign-functions=2`. Though this appears to be ignored (with no warnings from the compiler)

https://github.com/libretro/picodrive/blob/d6f625a1251c78caf6f2dc81c1ffdb724587bb24/Makefile#L26

Is there a hacky way to get this done with a modern build of emscripten?

(P.S. The "EMULATED_FUNCTION_POINTERS" option seems to have disappeared without a trace)

**Comment by kripken** [comment]
Yes, this looks like a clang bug. Testcase:
```cpp
void foo() {}
void bar() {}

int main(int argc, char **argv) {
  if (argc == 17) return (int)&foo;
  else return (int)&bar;
}
```
Compiling with or without `-falign-functions=2` makes no impact on the layout of the table. They are assigned adjacent indexes.

@sbc100 is this a bug in wasm-ld specifically perhaps, I think the table is laid out there?

**Reply by sbc100** [reply]
Yes, this would need to somehow be piped through to the linker. 

Regarding `EMULATED_FUNCTION_POINTERS` .. I don't think that was ever a setting, but we do have `EMULATE_FUNCTION_POINTER_CASTS` but that doesn't change anything about the absolute value of functions pointers.

**Reply by sbc100** [reply]
Interestingly the docs for `-falign-functions` says:  Enabled at levels -O2, -O3.   So I guess even if we added this it wouldn't work for `-O0` builds.. which seems odd.

**Reply by sbc100** [reply]
To get this working would likely require a fair amount work on the llvm side.  We would need to somehow encode the function alignment into each function the object file format.   I think depending on the bits of function pointer is not something you do in C/C++, so I'm inclined to say we simply don't support this.

**Reply by ethanaobrien** [reply]
>I'm inclined to say we simply don't support this.

Hm, that is disappointing. If I am working with an application that does expect functions to be aligned, do you know of any hacky ways to get them there? (As [for some reason](https://github.com/libretro/picodrive/blob/d6f625a1251c78caf6f2dc81c1ffdb724587bb24/Makefile#L26), its needed)

I also did some digging and I think the user in issue #6068 has/had the same need I do (though that setting is long since gone). There was a link to `emscripten-fastcomp` sent by @kripken (for what seems to be a solution), though that repo is long since gone.

I do also have reason to believe that this was possible back in `Oct 14 2018` due to a closed-source copy of this project supporting webassembly with the feature working - though they could just be doing something hacky to have gotten that working too. 

>Regarding `EMULATED_FUNCTION_POINTERS` .. I don't think that was ever a setting

I could possibly be incorrect, or it could be a setting in some other repo, but its mentioned [in the changelog under v1.34.6](https://github.com/emscripten-core/emscripten/blob/main/ChangeLog.md#v1346-8202015)

**Reply by sbc100** [reply]
Actually thinking about it, it might be possible to add a linker flag that would do this.   This wouldn't be the same compiler flag but should be easy to implement.   We would basically leave every other table slot empty.   

Would you like to try to write a wasm-ld patch maybe?

**Reply by ethanaobrien** [reply]
I can definitely take a look, yeah. Before I start digging through a codebase I’ve never seen, what should I be looking for?

**Reply by sbc100** [reply]
Function addresses in wasm are basically table indexes.  So you would want to add a new setting to wasm-ld that effects how this function works: https://github.com/llvm/llvm-project/blob/87dd31234a3c46ace097af73ecaedcdb6b306ef3/lld/wasm/SyntheticSections.cpp#L560-L568.

**Reply by ethanaobrien** [reply]
Ah, that does seem relatively simple.

I've modified the function to the below, assuming my logic is correct, it should align all functions by 2:
```C++
void ElemSection::addEntry(FunctionSymbol *sym) {
  if (sym->hasTableIndex() || sym->isStub)
    return;

  uint64_t alignment = 2;

  uint64_t index = config->tableBase + (indirectFunctions.size() * alignment);

  sym->setTableIndex(index);
  indirectFunctions.emplace_back(sym);
}
```

However, linking with `O2` or `O3` results in `wasm-opt` failing
```shell
wasm-opt: /build/emscripten/src/binaryen/src/mixed_arena.h:188: T& ArenaVectorBase<SubType, T>::operator[](size_t) const [with SubType = ArenaVector<wasm::Expression*>; T = wasm::Expression*; size_t = long unsigned int]: Assertion `index < usedElements' failed.
emcc: error: '/usr/bin/wasm-opt --strip-target-features --post-emscripten -O3 --low-memory-unused --asyncify --pass-arg=asyncify-propagate-addlist --pass-arg=asyncify-imports@env.invoke_*,env.__asyncjs__*,*.fd_sync,*.emscripten_promise_await,*.emscripten_idb_load,*.emscripten_idb_store,*.emscripten_idb_delete,*.emscripten_idb_exists,*.emscripten_idb_clear,*.emscripten_idb_load_blob,*.emscripten_idb_store_blob,*.emscripten_sleep,*.emscripten_wget_data,*.emscripten_scan_registers,*.emscripten_lazy_load_code,*._load_secondary_module,*.emscripten_fiber_swap,*.SDL_Delay --zero-filled-memory --pass-arg=directize-initial-contents-immutable picodrive_libretro.wasm -o picodrive_libretro.wasm --mvp-features --enable-multivalue --enable-mutable-globals --enable-reference-types --enable-sign-ext' failed (received SIGABRT (-6))
make: *** [Makefile.emscripten:236: picodrive_libretro.js] Error 1
```

Building with `O1` results in an `RuntimeError: index out of bounds` error. 

Perhaps it may not be as simple as modifying that function? Or my math is wrong

**Reply by sbc100** [reply]
I think likely the setTableIndex argument has to match the index into indirectFunctions.   Do you would likely need to push a one of more `nullptr` values onto `indirectFunctions` array.

**Reply by ethanaobrien** [reply]
Ah that would make sense. It's pretty unclean but this works.

```C++
void ElemSection::addEntry(FunctionSymbol *sym) {
  // Don't add stub functions to the wasm table.  The address of all stub
  // functions should be zero and they should they don't appear in the table.
  // They only exist so that the calls to missing functions can validate.
  uint32_t padding = 1;
  uint64_t alignment = 2;

  if (indirectFunctions.size() == 0 && padding > 0) {
    for (uint32_t i=0; i<padding; i++) {
      indirectFunctions.push_back(nullptr);
    }
  }

  uint64_t index = config->tableBase + indirectFunctions.size();

  if (sym->hasTableIndex() || sym->isStub)
    return;
  sym->setTableIndex(index);
  indirectFunctions.emplace_back(sym);

  if (alignment > 1) {
    for (uint32_t i=0; i<alignment-1; i++) {
      indirectFunctions.push_back(nullptr);
    }
  }
}
```

I also had mo modify `ElemSection::writeBody` to handle a `nullptr`

```C++

    if (sym == nullptr) {
      (void) tableIndex;
      writeUleb128(os, 0, "function index");
      ++tableIndex;
      continue;
    }
```

**Reply by sbc100** [reply]
Great!  Would you like to send a PR and we can discuss further there?

**Reply by ethanaobrien** [reply]
Opened here: https://github.com/llvm/llvm-project/pull/105532

Thanks!
